File: src/examples/go/generation/astra_cancellation.go
// ax-example:start
// title: Go Cancel Background Work
// group: generation
// description: Cancels a live Astra run through the high-level controller and observes cooperative tool cancellation.
// provider: openai
// env: OPENAI_API_KEY, OPENAI_APIKEY
// level: advanced
// order: 13
// ax-example:end
package main
import (
 "context"
 "fmt"
 "os"
 "strings"
 "sync/atomic"
 "time"
 ax "github.com/ax-llm/ax/packages/go"
)
func main() {
 key:=os.Getenv("OPENAI_API_KEY");if key=="" {key=os.Getenv("OPENAI_APIKEY")};if key=="" {panic("Set OPENAI_API_KEY or OPENAI_APIKEY.")}
 control:=ax.RunControl();settled:=make(chan struct{});var started atomic.Int64
 lookup:=ax.Fn("lookup").Execution("background").WithContextHandler(func(ctx context.Context,_ map[string]ax.Value)(ax.Value,error){started.Store(time.Now().UnixNano());control.Abort();select {case <-ctx.Done():close(settled);return "LATE: discard this result",nil;case <-time.After(2*time.Second):return nil,fmt.Errorf("Tool missed cancellation")}})
 program:=ax.NewAx("question -> answer",nil);program.Functions=[]ax.Tool{lookup}
 client:=ax.NewAI("openai",ax.Object("api_key",key,"model","gpt-6-astra","model_config",ax.Object("thinkingTokenBudget","low","max_tokens",2048)))
 ctx,cancel:=context.WithTimeout(context.Background(),90*time.Second);defer cancel()
 _,err:=program.Forward(ctx,client,ax.Object("question","Call lookup once and return its result."),ax.Object("control",control,"serviceTier","standard"))
 if err==nil||started.Load()==0||!strings.Contains(err.Error(),"unresolved calls") {panic(fmt.Sprintf("Expected cancelled pending work: %v",err))}
 elapsed:=time.Since(time.Unix(0,started.Load()));if elapsed>2*time.Second {panic("Cancellation blocked caller")};select {case <-settled:case <-time.After(2*time.Second):panic("Tool missed cancellation")}
 fmt.Printf("Cancelled in %s; %v\n",elapsed,err)
}

File: src/examples/cpp/generation/astra_cancellation.cpp
// ax-example:start
// title: C++ Cancel Background Work
// group: generation
// description: Cancels a live Astra run through the high-level controller and observes cooperative tool cancellation.
// provider: openai
// env: OPENAI_API_KEY, OPENAI_APIKEY
// level: advanced
// order: 13
// ax-example:end
#include "axllm/axllm.hpp"
#include <iostream>
using namespace axllm;
struct CancellationState {AxRunControl control;std::atomic<bool> settled{false};std::atomic<long long> started{0};};
int main(){
 const char* key=std::getenv("OPENAI_API_KEY");if(!key||!*key)key=std::getenv("OPENAI_APIKEY");if(!key||!*key)throw std::runtime_error("Set OPENAI_API_KEY or OPENAI_APIKEY.");
 auto state=std::make_shared<CancellationState>();
 Tool lookup("lookup","Look up the reference.");lookup.execution("background").context_handler([state](Value,const AxToolContext& context){state->started.store(std::chrono::steady_clock::now().time_since_epoch().count());state->control.abort();auto deadline=std::chrono::steady_clock::now()+std::chrono::seconds(2);while(!context.is_cancelled()&&std::chrono::steady_clock::now()<deadline)std::this_thread::sleep_for(std::chrono::milliseconds(1));if(!context.is_cancelled())throw std::runtime_error("Tool missed cancellation");state->settled.store(true);return Value("LATE: discard this result");});
 auto program=ax("question -> answer");program.add_tool(lookup);
 auto client=ai("openai",object({{"api_key",key},{"model","gpt-6-astra"},{"model_config",object({{"thinkingTokenBudget","low"},{"max_tokens",2048}})}}));
 try{program.forward(*client,object({{"question","Call lookup once and return its result."}}),object({{"control",state->control.value()},{"serviceTier","standard"}}));throw std::runtime_error("Cancelled run returned success");}
 catch(const AxError& error){auto started=std::chrono::steady_clock::time_point(std::chrono::steady_clock::duration(state->started.load()));auto elapsed=std::chrono::steady_clock::now()-started;if(!state->started.load()||std::string(error.what()).find("unresolved calls")==std::string::npos||elapsed>std::chrono::seconds(2))throw;auto deadline=std::chrono::steady_clock::now()+std::chrono::seconds(2);while(!state->settled.load()&&std::chrono::steady_clock::now()<deadline)std::this_thread::sleep_for(std::chrono::milliseconds(1));if(!state->settled.load())throw std::runtime_error("Tool missed cancellation");std::cout<<"Cancelled in "<<std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count()<<"ms; "<<error.what()<<"\n";}
}
